### PR TITLE
ENCD-5908-add-formatted-biorep-property-to-experiments-and-fcc-experiments

### PR DESCRIPTION
**Pull request by forresttanaka**
I put this new shared calculated property on the two objects that can have replicates: Experiment and FunctionalCharacterizationExperiment. Another option involved adding this property to the dataset object, so let me know if you feel that would work better.

### CODE REVIEW

File: src/encoded/types/shared_calculated_properties.py [hunk 1]
```diff
@@ -514,3 +514,20 @@ def replication_type(self, request, replicates=None, assay_term_name=None):
...
+class CalculatedBiologicalReplicates:
+    @calculated_property(schema={
+        "title": "Formatted biological replicates",
+        "description": "The biological replicate numbers associated with this dataset as formatted text.",
+        "comment": "Do not submit. This field is calculated through the experiment replicates.",
+        "type": "string"
+    })
+    def biological_replicates(self, request, replicates=None):
+        replicate_numbers = set()
+        for rep in replicates:
```
Comment (keenangraham) [inline src/encoded/types/shared_calculated_properties.py:528]: With `replicates=None` default value I think you could be trying to iterate over None here. Could maybe add `condition=replicates` to calculated property so it only runs if replicates is defined?

File: src/encoded/types/shared_calculated_properties.py [hunk 2]
```diff
@@ -514,3 +514,20 @@ def replication_type(self, request, replicates=None, assay_term_name=None):
...
+            replicateObject = request.embed(rep, '@@object')
```
Comment (keenangraham) [inline src/encoded/types/shared_calculated_properties.py]: Snake case.

File: src/encoded/tests/test_types_functional_characterization_experiment.py [hunk 3]
```diff
@@ -63,3 +63,24 @@ def test_fcc_crispr_assay_readout_method(testapp, functional_characterization_ex
...
+def test_fcc_biosample_replicates_1(testapp, functional_characterization_experiment_disruption_screen, biosample_1, library_1, replicate_1_fce, disruption_genetic_modification):
+    testapp.patch_json(biosample_1['@id'], {'genetic_modifications': [disruption_genetic_modification['@id']]})
+    testapp.patch_json(library_1['@id'], {'biosample': biosample_1['@id']})
+    testapp.patch_json(replicate_1_fce['@id'], {'library': library_1['@id']})
+    res = testapp.get(functional_characterization_experiment_disruption_screen['@id']+'@@index-data')
+    assert res.json['object']['biological_replicates'] == '1' 
+
+def test_fcc_biosample_replicates_0(testapp, functional_characterization_experiment_disruption_screen):
+    res = testapp.get(functional_characterization_experiment_disruption_screen['@id']+'@@index-data') 
+    assert not 'biological_replicates' in res.json['object']
```
Comment (keenangraham) [inline src/encoded/tests/test_types_functional_characterization_experiment.py]: I think `not in` a little more idiomatic: `assert 'biological_replicates' not in res.json['object']`

File: src/encoded/tests/test_types_experiment.py [hunk 4]
```diff
@@ -504,3 +504,32 @@ def test_experiment_default_analysis(
...
 
     assert res.json['object']['default_analysis'] == analysis_2['@id']
+
+def test_experiment_biosample_replicates_1_2(testapp, base_experiment, biosample_1, biosample_2, library_1, library_2, replicate_1_1, replicate_2_1):
+    testapp.patch_json(library_1['@id'], {'biosample': biosample_1['@id']})
+    testapp.patch_json(library_2['@id'], {'biosample': biosample_2['@id']})
+    testapp.patch_json(replicate_1_1['@id'], {'library': library_1['@id']})
+    testapp.patch_json(replicate_2_1['@id'], {'library': library_2['@id']})
+    testapp.patch_json(base_experiment['@id'], {'replicates': [replicate_1_1['@id'], replicate_2_1['@id']]})
+    res = testapp.get(base_experiment['@id']+'@@index-data')
```
Comment (keenangraham) [inline src/encoded/tests/test_types_experiment.py]: Space around `+`.

**Review by keenangraham — CHANGES_REQUESTED** [review]

File: src/encoded/types/shared_calculated_properties.py (hunk 1 above)
Comment (forresttanaka) [reply src/encoded/types/shared_calculated_properties.py:528]: Oh, _that’s_ what `condition` is for!

**Review by keenangraham — APPROVED** [review]

**Pull request merged**